### PR TITLE
Add "Quantization in Practice"  blogpost

### PR DESCRIPTION
**Pull request by suraj813**

### CODE REVIEW

**Comment by suraj813** [comment]
@woo-kim @holly1238 The MathJax sections aren't rendering correctly in the blog (they're too small + some formatting issues in one particular place). Do you have suggestions on how to fix this?

File: _posts/2022-1-19-quantization-in-practice.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.default_dynamic_qconfig}  # An empty key denotes the default applied to all modules
+model_prepared = quantize_fx.prepare_fx(m, qconfig_dict)
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Post-Training Static Quantization
+Similar to dynamic quantization, weights are pre-quantized. In dynamic, the activations are calibrated and quantized at inference time. In static, activations are precalibrated by passing sample inputs to the model. 
+
+This method has faster inference than dynamic quantization, but is less robust to accuracy drops from out-of-distribution inputs, i.e. if the model encounters data different from the sample inputs it has calibrated.
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:90]: A bit faster or much faster? Also above you recommend dynamic quantization for large models

And a lack of robustness against out of distribution inputs is a big deal, is this technique made for models that need to be fast but hardly change? At least based on what you describe it seems net worse than dynamic quantization

File: _posts/2022-1-19-quantization-in-practice.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+m = nn.Sequential(
+     nn.Conv1d(2,64,(8,)),
+     nn.ReLU(),
+     nn.Conv1d(64, 128, (1,)),
+     nn.ReLU()
+)
+
+## EAGER MODE
+"""Fuse"""
+torch.quantization.fuse_modules(m, ['0','1'], inplace=True) # fuse first Conv-ReLU pair
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:113]: cool feature I didn't know about! Why wouldn't someone fuse everything in this case, feels like this can be `for loop`ed to make people's life easier @jamesr66a 

Also briefly what does an inplace fusion mean?

File: _posts/2022-1-19-quantization-in-practice.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+"""Insert stubs"""
+m = nn.Sequential(torch.quantization.QuantStub(), 
+                  *m, 
+                  torch.quantization.DeQuantStub())
+
+"""Prepare"""
+m.qconfig = torch.quantization.get_default_qconfig('fbgemm')
+torch.quantization.prepare(m, inplace=True)
+
+"""Calibrate"""
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:125]: Is calibrating on random inputs helpful for static quantization if not I'd just add brief comment that this is a convenience thing

File: _posts/2022-1-19-quantization-in-practice.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    m(x)
+    
+"""Convert"""
+torch.quantization.convert(m, inplace=True)
+
+"""Check"""
+print(m[1].weight().element_size()) # 1 instead of 4 for FP32
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:135]: what does an element size of 4 mean?

File: _posts/2022-1-19-quantization-in-practice.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+
+## FX GRAPH
+from torch.quantization import quantize_fx
+qconfig_dict = {"": torch.quantization.get_default_qconfig('fbgemm')}
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:140]: what does fbgemm stand for? is this like saying `nn.Linear` in a `qconfig`?

File: _posts/2022-1-19-quantization-in-practice.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+# Calibrate 
+with torch.inference_mode():
+  for _ in range(10):
+    x = torch.rand(1,2,28)
+    model_prepared(x)
+# quantize
+model_quantized = quantize_fx.convert_fx(model_prepared)
+```
+
+### Quantization-aware Training (QAT)
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:153]: NVIDIA had a nice diagram on QAT since this flow is a bit complex would help to explain in a picture https://developer.nvidia.com/blog/improving-int8-accuracy-using-quantization-aware-training-and-tao-toolkit/

Same comment for the other 2 techniques

File: _posts/2022-1-19-quantization-in-practice.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+It's likely that you can still use QAT by "fine-tuning" it on a sample of the training dataset, but I did not try it on demucs (yet).
+
+
+## Quantizing "real-world" models
+
+**Download the [notebook](https://gist.github.com/suraj813/735357e56321237950a0348b50f2f3b4) or run it on [Colab](https://colab.research.google.com/gist/suraj813/735357e56321237950a0348b50f2f3b4/fx-and-eager-mode-quantization-example.ipynb) (note that Colab runtimes may differ significantly from local machines).**
+
+Traceable models can be easily quantized with FX Graph Mode, but it's possible the model you're using is not traceable end-to-end. Maybe it has loops or `if` statements on inputs (dynamic control flow), or relies on third-party libraries. The model I use in this example has [dynamic control flow and uses third-party libraries](https://github.com/facebookresearch/demucs/blob/v2/demucs/model.py). As a result, it cannot be symbolically traced directly. In this code walkthrough, I show how you can bypass this limitation by quantizing the child modules individually for FX Graph Mode, and how to patch Quant/DeQuant stubs in Eager Mode.
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:174]: The model `we use` since this is an official tutorial

File: _posts/2022-1-19-quantization-in-practice.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+
+Figure 6(a) shows why. In PTQ, the FP32 model's parameters are chosen by optimizing on the FP32 loss, and then projected to INT8 space. Depending on where the new INT8 weights lie on the loss curve, model accuracies can significantly change.
+
+In QAT, the FP32 parameters are chosen by also optimizing on the INT8 loss. This allows the model to identify a wider region in the loss function (Figure 6(b)), and identify FP32 parameters such that quantizing them does not significantly affect accuracy.
+
+It's likely that you can still use QAT by "fine-tuning" it on a sample of the training dataset, but I did not try it on demucs (yet).
+
+
+## Quantizing "real-world" models
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:170]: This is interesting in its own right, I'd rather you either write it out in this blog post or do a follow up post on quantizing non tracable models

File: _posts/2022-1-19-quantization-in-practice.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+**Download the [notebook](https://gist.github.com/suraj813/735357e56321237950a0348b50f2f3b4) or run it on [Colab](https://colab.research.google.com/gist/suraj813/735357e56321237950a0348b50f2f3b4/fx-and-eager-mode-quantization-example.ipynb) (note that Colab runtimes may differ significantly from local machines).**
+
+Traceable models can be easily quantized with FX Graph Mode, but it's possible the model you're using is not traceable end-to-end. Maybe it has loops or `if` statements on inputs (dynamic control flow), or relies on third-party libraries. The model I use in this example has [dynamic control flow and uses third-party libraries](https://github.com/facebookresearch/demucs/blob/v2/demucs/model.py). As a result, it cannot be symbolically traced directly. In this code walkthrough, I show how you can bypass this limitation by quantizing the child modules individually for FX Graph Mode, and how to patch Quant/DeQuant stubs in Eager Mode.
+
+
+
+## What's next - Define-by-Run Quantization
```
Comment (msaroufim) [inline _posts/2022-1-19-quantization-in-practice.md:178]: This would be in the followup post too and you can basically gett people excited about reading this followup post by describing briefly how cool it is and how it works

Also make sure to synthesize everything people learnt at the very end, that your table and figures will help this doc be a reference for people new to quantization

File: _posts/2022-1-19-quantization-in-practice.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+> If someone asks you what time it is, you don't respond "10:14:34:430705", but you might say "a quarter past 10".
+
+Quantization has roots in information compression; in deep networks it refers to reducing the numerical precision of its weights and/or activations. Overparameterized DNNs have more degrees of freedom and this makes them good candidates for information compression. When you quantize a model, two things generally happen - the model gets smaller and runs with better efficiency. Processing 8-bit numbers is faster than 32-bit numbers, and a smaller model has lower memory footprint and power consumption.
+
+At the heart of it all is a **mapping function**, a linear projection from floating-point to integer space: $Q(r) = round(r/S + Z)$
+
+To reconvert to floating point space, the inverse function is given by $\tilde r = (Q(r) - Z) \cdot S$. $\tilde r \neq r$, and their difference constitutes the **quantization error**.
+
+The scaling factor $S$ is simply the ratio of the input range to the output range: $S = \frac{\beta - \alpha}{\beta_q - \alpha_q}$
```
[the remark this answers is not in the document]
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:24]: A) to provide an intuition of where this number is coming from, and B) to help make sense of how a/symmetric quantization schemes are better/worse for a given use case

File: _posts/2022-1-19-quantization-in-practice.md [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+where [$\alpha, \beta$] is the clipping range of the input, i.e. the boundaries of permissible inputs. [$\alpha_q, \beta_q$] is the range in quantized output space that it is mapped to. For 8-bit quantization, the output range $\beta_q - \alpha_q <= (2^8 - 1) $.
+
+The process of choosing the appropriate input range is known as **calibration**; commonly used methods are MinMax and Entropy.
```
[the remark this answers is not in the document]
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:27]: it varies depending on the quantization scheme chosen. I'll make this more obvious here

Comment (jerryzh168) [inline _posts/2022-1-19-quantization-in-practice.md:27]: should we mention observers here? calibration is just a step that runs some sample data through the model and also through the inserted observers, so that the values can be recorded and used to calculate quantization parameters

File: _posts/2022-1-19-quantization-in-practice.md [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+$-\alpha = \beta = max(|max(r)|,|min(r)|)$. This removes the need of a zero-point offset in the mapping function. Asymmetric or **affine** schemes simply assign the boundaries to the minimum and maximum observed values. Asymmetric schemes have a tighter clipping range (for non-negative ReLU activations, for instance) but require a non-zero offset.
+
+
+### PyTorch Classes
+`Observer` modules ([docs](https://PyTorch.org/docs/stable/torch.quantization.html?highlight=observer#observers), [code](https://github.com/PyTorch/PyTorch/blob/748d9d24940cd17938df963456c90fa1a13f3932/torch/ao/quantization/observer.py#L88)) collect statistics on the input values and calculate the qparams $S, Z$. 
+
+The `QConfig` ([code](https://github.com/PyTorch/PyTorch/blob/d6b15bfcbdaff8eb73fa750ee47cef4ccee1cd92/torch/ao/quantization/qconfig.py#L165)) NamedTuple specifies the observers and quantization schemes for the network's weights and activations. The default qconfig is at `torch.quantization.get_default_qconfig(backend)` where `backend='fbgemm'` for x86 CPU and `backend='qnnpack'` for ARM. 
+
+
+## In PyTorch
```
Comment (jerryzh168) [inline _posts/2022-1-19-quantization-in-practice.md:45]: there are two other dimensions, quantization mode: static/dynamic/weight only, and backend: server cpu/mobile cpu, gpu. I have more info in my slides

File: _posts/2022-1-19-quantization-in-practice.md [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+- if $S, Z$ are computed during, or after training (*quantization-aware training* v/s *post-training quantization*)
+
+Each approach has its unique tradeoffs; for instance FX Graph Mode can automagically figure out the right quantization configurations, but only for models that are [symbolically traceable](https://PyTorch.org/docs/stable/fx.html#torch.fx.symbolic_trace). Dynamic quantization can offer better precision at the cost of additional overhead in each inference. 
+
+### Post-Training Dynamic Quantization 
+The model's weights are pre-quantized before inference, but the activations are calibrated and quantized on the fly during inference. The simplest of all approaches, it has a one line API call in `torch.quantization.quantize_dynamic`
+
+Because the calibrations are bespoke, clipping ranges can be tighter; dynamic quantization can therefore theoretically give higher accuracies, but calibrating and quantizing each layer's activations can add to the overhead. 
+
+For this reason, it is best suited for models where most of the execution time is spent in loading weights from memory (think very large models with billions of parameters).
```
[the remark this answers is not in the document]
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:59]: what i was trying to convey is that this is better-suited for models where throughput is dominated by memory access (vs compute). I'm basing it off this blogpost https://pytorch.org/blog/introduction-to-quantization-on-pytorch/#choosing-an-approach. I'll rephrase this section to make the distinction clearer

[the remark this answers is not in the document]
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:59]: what i was trying to convey is that this is better-suited for models where throughput is dominated by memory access (vs compute). I'm basing it off this blogpost https://pytorch.org/blog/introduction-to-quantization-on-pytorch/#choosing-an-approach. I'll rephrase this section to make the distinction clearer

File: _posts/2022-1-19-quantization-in-practice.md (hunk 1 above)
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:90]: the first question is a bit tricky to answer because that can depend on the model architecture. generally, this is recommended for models that have large activations bottlenecking compute and memory access eg: CNNs

yes, it is more susceptible to accuracy issues from drift, possibly ameliorated by regular re-calibration. i'll make this point clearer. thanks!

File: _posts/2022-1-19-quantization-in-practice.md (hunk 3 above)
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:125]: Good callout, I thought I had a comment in the code about this somewhere. Thanks!

File: _posts/2022-1-19-quantization-in-practice.md (hunk 4 above)
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:135]: units in bytes, added detail

File: _posts/2022-1-19-quantization-in-practice.md (hunk 8 above)
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:170]: Yeah it's the most interesting part of the article imo. But it's also very verbose - containing code and commentary - which is why I've only linked to the notebook from here.  A follow-up post might be easier to parse, thanks for the suggestion.

File: _posts/2022-1-19-quantization-in-practice.md (hunk 12 above)
Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:45]: are the gpu backends usable yet? i don't see this as a backend in 
- https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/qconfig.py 
-  https://github.com/pytorch/pytorch/blob/c371542efc31b1abfe6f388042aa3ab0cef935f2/torch/backends/quantized/__init__.py

Comment (jerryzh168) [reply _posts/2022-1-19-quantization-in-practice.md]: GPU path with TensorRT is in prototype for internal users, GPU path with cudnn and cuda is WIP

File: _posts/2022-1-19-quantization-in-practice.md [hunk 14]
```diff
@@ -0,0 +1,292 @@
...
+featured-img: ''
+---
+
+There are a few different ways to quantize your model with PyTorch. In this blog post, we'll take a look at how each technique looks like in practice. I will use a non-standard model that is not traceable, to paint an accurate picture of how much effort is really needed when quantizing your model.
+
+<div class="text-center">
+  <img src="/assets/images/quantization_gif.gif" width="60%">
+</div>
+
+## A quick introduction to quantization
```
Comment (jerryzh168) [inline _posts/2022-1-19-quantization-in-practice.md:14]: would you like to integrate some part of this in the official quantization api?

Comment (suraj813) [reply _posts/2022-1-19-quantization-in-practice.md:16]: do you mean the API itself, or in the docs? I'm happy to add it!

Comment (jerryzh168) [reply _posts/2022-1-19-quantization-in-practice.md:16]: the documentation, we can add the explanations of core things like qconfig, quantized tensor, observer/fake_quant, qscheme to documentation: https://pytorch.org/docs/master/quantization.html#
